### PR TITLE
Add root-cause to the e2e tests

### PR DESCRIPTION
**Pull request by Bnaya**
Hey 👋

We wrote an open source tool called “Root Cause” that gives a better debugging experience for puppeteer/playwright tests. You can think of it as “cypress like debugging for puppeteer and playwright”. 

When integrated, Root Cause will trace calls (or “test steps”) to the test automation library (puppeteer/playwright) and capture screenshot, console and network logs and other debug info, and make it all viewable in an intuitive viewer.

<p align="center"></p>

<p align="center"></p>

I was looking for open-source libraries with nice puppeteer/playwright e2e tests to integrate with so we can gather product feedback.

This PR adds root cause to this project. Note none of this is automatic and I actually went through the code to integrate.

After tests run with Root Cause integrated, you can see the results using the cli:
npx root-cause ls / npx root-cause show or yarn’s counterpart yarn root-cause ls/show

https://help.testim.io/docs/viewing-local-results-root-cause-cli

Cloud Viewer

Testim.io also has a cloud offering of root-cause (like circleci for open source). It’s still very early on but we’d like to offer you “root cause cloud” for free. It basically makes it much easier to collaborate on e2e test results and debug why e2e tests failed.

I didn’t want to go ahead and register for you because that felt pushy to me and I’m trying to be helpful and gather feedback for this thing we’ve built. 

To upload results to the cloud viewer, you will need to register to testim at https://app.testim.io/#/signup?fromRootCause=true and add your access token as an environment variable as described in the getting started page. Once you do that please let me know and I’ll be happy to enable it or your account.

Cheers,
Bnaya

### CODE REVIEW

**Comment by mskelton** [comment]
Wow, thanks for the PR!  Another open source project I created which uses Playwright is https://github.com/mskelton/prop-types-to-ts which I might want to give this a try with.  Before merging I'm going to pull this locally and try it out so I can give some feedback on how it works, but overall it looks like a really awesome product, so great work!

File: e2e/utils/url.js [hunk 1]
```diff
@@ -2,4 +2,4 @@ exports.url =
   process.env.BASE_URL ||
   (process.env.NODE_ENV === "production"
     ? "https://nt-challenge.web.app"
-    : "http://localhost:8100");
+    : "http://localhost:3000");
```
Comment (mskelton) [inline e2e/utils/url.js:5]: Why was this changed?

File: package.json [hunk 2]
```diff
@@ -48,6 +48,7 @@
   },
   "devDependencies": {
     "@capacitor/cli": "2.0.0",
+    "@testim/root-cause-jest": "^0.1.23",
```
Comment (mskelton) [inline package.json:51]: This can probably be removed in favor of the e2e package.json

**Review by mskelton — COMMENTED** [review]
I've been trying this out a bit and I had some feedback I wanted to share:

1. The boilerplate for the `runId` feels pretty odd.  I wish that came out of the box.
1. Opening the results is kind of a pain.  Having to copy and paste the npx command is quite annoying and I wish that I could just click a url and have it load the results.  Maybe it could be a file path in the node_modules folder or something similar where it stores the results so it doesn't require running a server?

File: e2e/jest.config.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
 module.exports = {
   preset: "jest-playwright-preset",
-  setupFilesAfterEnv: ["expect-playwright"],
```
Comment (Bnaya) [inline e2e/jest.config.js:3]: The updated `jest-playwright-preset` include also `expect-playwright`

File: e2e/jest.config.js [hunk 4]
```diff
@@ -1,4 +1,5 @@
...
+  testEnvironment: '@testim/root-cause-jest/lib/RootCauseJestEnv',
```
Comment (Bnaya) [inline e2e/jest.config.js:3]: We've revamp our jest integration to be based on custom env (that infers and load the actual env under the hood)
Now the runId is not mandatory

**Comment by Bnaya** [comment]
I've made some changes to the PR,
I had to upgrade your playwright related packages, i hope it's ok.

> The boilerplate for the runId feels pretty odd. I wish that came out of the box.

We've update the integration and now it's now needed. Due to how jest works, passing explicit runId is the most "compatible way" but we've managed to work around is in a way that will work for most cases.

> Opening the results is kind of a pain. Having to copy and paste the npx command is quite annoying and I wish that I could just click a url and have it load the results. Maybe it could be a file path in the node_modules folder or something similar where it stores the results so it doesn't require running a server?

So you would prefer the result to be a static HTML file that can be simply open?
We were evaluating this option, but some of the features will are working on will not work that way.
We will give it another look.

Additionally You may also use `npx root-cause ls`,
or  `npx root-cause show`that will prompted you to select result

**prop-types-to-ts**
I'm also on it

Thanks!

**Comment by mskelton** [comment]
> So you would prefer the result to be a static HTML file that can be simply open?

Yeah, I think that would be easiest for the user.  Additionally, this would allow uploading these artifacts to CI like Buildkite which can upload a specific folder path which can then be viewed in the Buildkite artifacts tab.  Not sure how other CI systems work, but just a thought.

> Additionally You may also use npx root-cause ls,

I tried that, but it is confusing which run is which.  Additionally, having to switch between running jest and `npx root-cause ls` is quite annoying when running tests.  Interested to know why certain features wouldn't work with the static HTML.

File: e2e/.root-causerc.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
+    "features": {
+        "jestAssertions": true
```
Comment (mskelton) [inline e2e/.root-causerc.json:3]: What does this do?

**Review by mskelton — APPROVED** [review]
Nice!

File: e2e/.root-causerc.json (hunk 5 above)
Comment (Bnaya) [reply e2e/.root-causerc.json:3]: The config file is documented here,
https://help.testim.io/docs/root-cause-configuration-file-root-causerc

But basically when `jestAssertions` is on, we track also calls to `expect(...).toBe(...)`  and record them as steps from the tests, and not only playwright calls.
Additionally, is case of matches that makes internal calls to playwright apis (eg `expect(page).toHaveText('[data-testid="title"]', "NT Challenge");`)
We will group the internal calls into the expect step call so it will be more clear.

**Comment by mskelton** [comment]
Thanks!
